### ISSUE TITLE
trailing slash in parameter vs non

### ISSUE THREAD

**Issue by mwilson8** [tiangolo/fastapi#1127]
### First check

* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.

### Description

Is there a way to have FastAPI treat a parameter the same if it does or does not have a trailing `/` ?

### Additional context

I want to replicate this in a more concise manner and/or configure FastAPI so I don't have to do this for each path similarly

```
@app.post("/extract")
@app.post("/extract/")
async def create_extraction(...):
```
EDIT: simplified example code

**Comment by jmriebold** [comment]
What version of fastapi are you using? As of [0.50.0](https://github.com/tiangolo/fastapi/releases/tag/0.50.0) the trailing slash should be automatically stripped now that fastapi's using the latest version of starlette.

Also for what it's worth, even without that feature you could always do something like this:

```python
@app.post("/extract/?")
async def create_extraction(...):
```

since routes can take regexes.

**Comment by mwilson8** [comment]
@jmriebold we're unfortunately pinned on version 0.47.1  
I wasn't aware that FastAPI supported regex. Awesome! Thanks!

**Comment by tiangolo** [comment]
Thanks for the help @jmriebold ! :bowing_man: 

Thanks for reporting back and closing the issue @mwilson8 :+1:

**Comment by falkben** [comment]
FYI, @jmriebold I don't think path routes can take regex's anymore like this.

Since Starlette version 0.13.5 regex is escaped.  See the PR here: https://github.com/encode/starlette/pull/932 and further discussion here: https://github.com/encode/starlette/pull/1010

also... in my testing, Starlette is flexible about the trailing slash (and i think by extension fastapi) so you shouldn't need the trailing `?`.

**Comment by jmriebold** [comment]
Yeah, I noticed that too. Appears to be the case that regexes were never intentionally supported, it was just an artifact of how paths were constructed.

**Comment by falkben** [comment]
It seems like the best option is to just allow Starlette to do the redirect for you, so long as you are okay with redirects.  Note, that curl usually doesn't automatically redirect.  And neither does requests for POSTs.  So there are some valid reasons to not want your users to have to deal with redirects if you have routes they are using programmatically.

If you don't want the redirect, I think the other option is using multiple decorators as suggested in the original question.

I wish Starlette didn't perform a redirect but just directly accessed the other route, if it matched it. There's probably a good reason they don't do this, though.

**Comment by jmriebold** [comment]
Yeah I personally thought regexes were an elegant way to do it, but have since switched to an explicit redirect route.

**Comment by ErhoSen** [comment]
I think this Starlette release broke a lot of services (including mine), and forces us to choose between ugly
```python
@router.get("/block", response_model=pydantic_models.BlockResponse)
@router.get("/block/", response_model=pydantic_models.BlockResponse)
async def blocks(request: Request):
    ...
```

And ugly
```python
def slash_patcher(app):
    app.router.redirect_slashes = False
    routes = copy.deepcopy(app.router.routes)
    for r in routes:
        if r.path.endswith("/"):
            app.add_route(r.path.rstrip("/"), ...)
        else:
            app.add_route(r.path + "/", ...)
```

Maybe we can fix it somehow in general way?

**Comment by falkben** [comment]
Thanks for sharing your slash patcher. 

Perhaps a pull request or new issue would get the ball rolling?

**Comment by ErhoSen** [comment]
I don't thinks this is a good solution: for example `/docs` become a mess:
![image](https://user-images.githubusercontent.com/5013076/92586837-767f5680-f29f-11ea-87a7-00cde5969f79.png)

**Comment by falkben** [comment]
Yea, pretty gross.  I also think this is probably easier handled in Starlette

**Comment by ErhoSen** [comment]
In our team we decided to patch proxy-server for all our FastAPI applications: it ensures that every request does not contain `\` at the end of path. (otherwise it just removes slashes) :|

**Comment by Coppini** [comment]
Are there any updates on this? Would it be possible to support regex or multiple route options like in @ErhoSen suggestion
> ```python
> @router.get("/block", response_model=pydantic_models.BlockResponse)
> @router.get("/block/", response_model=pydantic_models.BlockResponse)
> async def blocks(request: Request):
>     ...
> ```

Without ending up with the `/docs` mess he later exemplified in the screenshot?

I've just encountered the same problem, and either using REGEX or being able to add multiple route options would be really helpful.

**Comment by falkben** [comment]
In general, redirects should occur automatically to the endpoint w/ or w/o the trailing slash (depending on which one you have configured.

If it's absolutely needed to have both routes, you can specify certain routes to *not* be included in the docs with  `include_in_schema=False`

This gets asked frequently. See this issue for some additional solutions: https://github.com/tiangolo/fastapi/issues/2060

**Comment by Akkarine** [comment]
You should use trailing slash, because there is caveat, if you have other endpoints with same path, like:

`GET /v1/users/{user_id}`
`GET /v1/users/balance`

On calls for balance, `/v1/users/{user_id}` could be called with `user_id=balance`
